intellij/src/saros/intellij/ui/util/UIActiveProjectProvider.java

```diff
@@ -0,0 +1,79 @@
+package saros.intellij.ui.util;
+
+import com.intellij.ide.DataManager;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.actionSystem.DataKeys;
+import com.intellij.openapi.project.Project;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.concurrency.Promise;
+import saros.session.ISarosSession;
+import saros.session.ISarosSessionManager;
+import saros.session.ISessionLifecycleListener;
+import saros.session.SessionEndReason;
+
+/**
+ * Class to provide access to an active project object. This class should only be used for UI
+ * purposes.
+ */
+public class UIActiveProjectProvider {
+  private static final Logger log = Logger.getLogger(UIActiveProjectProvider.class);
+
+  private static final int PROJECT_REQUEST_TIMEOUT = 1000;
+
+  private volatile ISarosSession sarosSession;
+
+  @SuppressWarnings("FieldCanBeLocal")
+  private ISessionLifecycleListener sessionLifecycleListener =
+      new ISessionLifecycleListener() {
+        @Override
+        public void sessionStarted(ISarosSession session) {
+          sarosSession = session;
+        }
+
+        @Override
+        public void sessionEnded(ISarosSession session, SessionEndReason reason) {
+          sarosSession = null;
+        }
+      };
+
+  public UIActiveProjectProvider(ISarosSessionManager sarosSessionManager) {
+    sarosSessionManager.addSessionLifecycleListener(sessionLifecycleListener);
+  }
+
+  /**
+   * Returns a current project object.
+   *
+   * <p>If there is currently a session, the session project will be returned. Otherwise, the
+   * currently focused project will be requested and returned. In this case, <code>null</code> is
+   * possible.
+   *
+   * <p><b>NOTE:</b> The project returned by this method should only be used to for UI purposes.
```

Comment:
I guess this method should be similar to its Eclipse counterpart https://github.com/saros-project/saros/blob/master/eclipse/src/saros/ui/util/SWTUtils.java#L202

But I do not understand

1. The Saros Part (I mean it is for UI purposes, so why do fetch it from the session) ?
2. Doesn't the getDataContextFromFocusAsync always produces a timeout when called from the Swing/AWT dispatch thread ?

Reply:
1. In an upcoming change, I am going to move the project object (probably contained in a wrapper) to the session context. The held project will be initialized as part of the negotiation, where the user chooses a project module combination to share (instead just a module).
I would prefer to use this project object for UI purposes to ensure that the notification/dialogs are displayed in the project that is actually shared and not the focused project. (This should not matter for dialogs but is should matter for notifications as those are only shown in the specific project.)
2. I am not sure. It does seem to work in general for the UI components, but I have only done minor testing. But it's a good hint. Will look into it.

Reply:
Yes, you were absolutely right. This method always times out when called from the AWT event dispatcher thread. The method internally calls [`doWhenFocusSettlesDown`](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/openapi/wm/impl/FocusManagerImpl.java#L178), which executes asynchronously on the AWT event dispatcher thread.

Will move the project requests out of the dispatcher thread blocks and add an assertion ensuring that the method is not called from the dispatcher thread.

Reply:
Ok, that does not work. Will have to think about this more.